truffle/src/main/java/org/jruby/truffle/core/format/printf/PrintfSimpleParser.java

```diff
@@ -0,0 +1,592 @@
+        }
+        if (argType == ArgType.NUMBERED) {
+            throw new InvalidFormatException("named" + new String(name) + " after numbered");
+        }
+    }
+
+    private enum ArgType {
+        NONE,
+        NUMBERED,
+        UNNUMBERED,
+        NAMED
+    }
+
+    private void checkPosArg(int relativeArgumentIndex, int absoluteArgumentIndex) {
+        if (relativeArgumentIndex > 0) {
+            throw new InvalidFormatException("numbered(" + absoluteArgumentIndex + ") after unnumbered(" + relativeArgumentIndex + ")");
+        }
+        if (relativeArgumentIndex == -2) {
+            throw new InvalidFormatException("numbered(" + absoluteArgumentIndex + ") after named");
+        }
+        if (absoluteArgumentIndex < 1) {
+            throw new InvalidFormatException("invalid index - " + absoluteArgumentIndex + "$");
+        }
+    }
+
+    public LookAheadResult getNum(int startI, int end) {
+        StringBuilder sb = new StringBuilder();
+
+        int moreChars = 0;
+        for (int i = startI; i < end; i++) {
+            char nextChar = source[i];
+            if (!isDigit(nextChar)) {
+                break;
+            } else {
+                sb.append(nextChar);
+                moreChars += 1;
+            }
+        }
+
+        final int nextI = startI + moreChars;
+
+        if (nextI >= end) {
+            throw new InvalidFormatException("malformed format string - %%*[0-9]");
+        }
+
+        Integer result;
+        if (sb.length() > 0) {
+            result = Integer.parseInt(sb.toString());
+        } else {
+            result = null;
+        }
+        return new LookAheadResult(result, nextI);
+    }
+
+    public LookAheadResult getNumberDollar(int startI, int end) {
+        LookAheadResult lar = getNum(startI, end);
+        Integer result = null;
+        int newI = startI;
+        if (lar.getNumber() != null) {
+            final int nextI = lar.getNextI();
+            if (valueAt(nextI) != null && valueAt(nextI) == '$') {
+                result = lar.getNumber();
+                newI = nextI + 1;
+                if (result < 1) {
+                    throw new InvalidFormatException("invalid index - " + result + "$");
+                }
+            }
+        }
+        return new LookAheadResult(result, newI);
+    }
+
+    public static class LookAheadResult {
+        private Integer number;
+        private int nextI;
+
+        public LookAheadResult(Integer number, int nextI) {
+            this.number = number;
+            this.nextI = nextI;
+        }
+
+        public Integer getNumber() {
+            return number;
+        }
+
+        public int getNextI() {
+            return nextI;
+        }
+    }
+
+    public static boolean isDigit(char c) {
+        return c == '0' || c == '1' || c == '2' || c == '3' || c == '4' || c == '5' || c == '6' || c == '7' || c == '8' || c == '9';
+    }
+
+    public static class SprintfConfig {
```

Comment:
If it's some kind of private implementation details and relatively small I think static nested classes are fine.
But here it's neither so I would move it outside as well.